### PR TITLE
kubelet/rkt: Add SyncPod() to rkt.

### PR DESCRIPTION
**Pull request by yifan-gu**
@vmarmol @yujuhong @dchen1107 @jonboulle @bakins

Currently, if any containers in the pod need to be restarted, the whole pod will be restarted.

WIP, need to move `shouldContainerBeRestarted` to hooks so we can reuse the code.

### CODE REVIEW

File: pkg/kubelet/rkt/sync.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/golang/glog"
+)
+
+// SyncPod makes
+func (r *Runtime) SyncPod(runningPod kubecontainer.Pod, pod *api.Pod, podStatus api.PodStatus, podVolumes map[string]volume.Volume) error {
+	podFullName := kubecontainer.GetPodFullName(pod)
+	uid := pod.UID
+
+	if len(runningPod.Containers) == 0 {
+		glog.V(4).Infof("Pod is not running, start it")
```
Comment (xiang90) [inline pkg/kubelet/rkt/sync.go]: does the running pod has an identity/description at this point? The log is very vague.

File: pkg/kubelet/rkt/sync.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import (
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kubecontainer "github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/container"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/probe"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume"
+	"github.com/golang/glog"
+)
+
+// SyncPod makes
```
Comment (xiang90) [inline pkg/kubelet/rkt/sync.go]: ?

File: pkg/kubelet/rkt/sync.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			delete(unidentifiedContainers, c.ID)
+			continue
+		}
+		if result == probe.Success {
+			glog.V(4).Infof("Probe success: %q", container.Name)
+			delete(unidentifiedContainers, c.ID)
+			continue
+		}
+		glog.Infof("Pod %q container %q is unhealthy (probe result: %v), it will be killed and re-created.", podFullName, container.Name, result)
+		restartPod = true
```
Comment (xiang90) [inline pkg/kubelet/rkt/sync.go]: shall we also break out in this case? Actually i think we should follow the previous logic here. If ...{ restartPod = true; break}; normal logic

**Comment by yifan-gu** [comment]
Need work around pod volumes.

File: pkg/kubelet/rkt/rkt.go [hunk 4]
```diff
@@ -69,9 +70,12 @@ const (
...
 // on the machine.
 type Runtime struct {
-	systemd *dbus.Conn
-	absPath string
-	config  *Config
+	generator        kubecontainer.RunContainerOptionsGenerator
+	readinessManager *kubecontainer.ReadinessManager
+	prober           prober.Prober
+	systemd          *dbus.Conn
+	absPath          string
```
Comment (vmarmol) [inline pkg/kubelet/rkt/rkt.go]: nit: I know this was here before, but can we document absPath?

**Comment by vmarmol** [comment]
Looks good overall, couple of things to node

- Missing restarting if pod spec changed: ports and host networking mode (can be a TODO)
- Add a TODO for supporting network plugins (we need to figure out how they'd work in rkt)

What's missing from the volume stuff?

**Comment by yifan-gu** [comment]
@vmarmol Thanks! 
I think ports and host networking changes will change the hash, no?

Volumes are required for launching a pod since now Im not using `RunContainerOptions`. Hmm, I am just going to bypass the volumes, and figure out how to inject RunContainerOptionsGenerator in rkt after this PR, opinions?

File: pkg/kubelet/rkt/rkt.go (hunk 4 above)
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: Sure, the name is confusing actually :(

**Comment by vmarmol** [comment]
The ports yes, the network mode no since it is as the pod level.

I'm fine doing the volumes later. You're already injecting the generator can you use that?

**Comment by yifan-gu** [comment]
@vmarmol That requires some modification to the RunPod(), I will figure out.

**Comment by yujuhong** [comment]
@vmarmol, @yifan-gu, do we still want to keep `RunPod` in the interface? Kubelet doesn't distinguish between a new and an existing pod, so it calls only `SyncPod` anyway...

`DockerManager` does not have`RunPod` and this is last function blocking it from implementing the `Runtime` interface. I can put a placeholder there if we want to keep this function, or I can just delete it :)

**Comment by yifan-gu** [comment]
@yujuhong We can delete it I think! Thanks for the reminder

**Comment by yifan-gu** [comment]
Have kick the volumes from the rkt's syncPod()'s signature. Will use RunContainerOptions.
But I think we can merge this once green and add switches for kubelet to choose docker/rkt next.

There are pumps in rkt's syncPod now, which I will figure out today or tomorrow.

@vmarmol @yujuhong

**Comment by vmarmol** [comment]
+1 to removing @yujuhong 

@yifan-gu we (@yujuhong or I) can work on the switch in Kubelet to free you up for rkt support (unless you've already started)

**Comment by yifan-gu** [comment]
Thanks @vmarmol !

**Comment by vmarmol** [comment]
LGTM, thanks @yifan-gu! Will start getting the Kubelet switch in. @yujuhong is working on the Runtime interface.

**Pull request merged**